lms/djangoapps/course_api/blocks/transformers/milestones.py

```diff
@@ -172,3 +168,40 @@ def gated_by_required_content(block_key, block_structure, required_content):  #
             return True
 
         return False
+
+    def _generate_special_exam_attempt_context(self, block_key, block_structure, usage_info):
+        """
+        Helper method which generates the special exam attempt context.
+        Either calls into proctoring or, if exams ida waffle flag on, then get internally.
+        """
+        special_exam_attempt_context = None
+
+        # if exams waffle flag enabled, get exam type internally
+        if exams_ida_enabled(block_key.course_key):
```

Comment:
One thing that's a little tricky with creating a function that both can use is that the parameters are different (since milestones is using block structures, and special exams is using sequences). So even though the code is really similar, there are key differences in the input and specifically the way that each method is getting the exam type.

Reply:
ugh, I just read the other version of this and it's annoyingly similar but every piece of data is in a different place

combining them would require like seven arguments. But they might be clearer arguments:

user_id, course_key, block_key (or is this usage_key or location_key?), is_practice, is_proctored, is_timed,

Reply:
Haha, yeah exactly! Sounds good, just wasn't sure if there was a limit to the number of args haha.

Reply:
I mean in general you want to keep args down because it gets confusing.

But args which are a list of very simple data items are less concerning than fancy args - if your function takes six heavy objects it could access any field from them, mutate them, etc. If it takes six strings, there is only so fancy it can be.